src/main/java/net/imglib2/img/ImgFactory.java

```diff
@@ -166,4 +148,84 @@
 	{
 		return imgFactory( typeSupplier.get() );
 	}
+
+
+	/*
+	 * -----------------------------------------------------------------------
+	 *
+	 * Deprecated API.
+	 *
+	 * Supports backwards compatibility with ImgFactories that are constructed
+	 * without a type instance or supplier.
+	 *
+	 * -----------------------------------------------------------------------
+	 */
+
+	@Deprecated
+	public ImgFactory() {
+		t = null;
+		supplier = null;
+	}
+
+	@Deprecated
+	public abstract Img< T > create( final long[] dim, final T type );
+
+	@Deprecated
+	public Img< T > create( final Dimensions dim, final T type )
+	{
+		final long[] size = new long[ dim.numDimensions() ];
+		dim.dimensions( size );
+
+		return create( size, cache( type ) );
+	}
+
+	@Deprecated
+	public Img< T > create( final int[] dim, final T type )
+	{
+		return create( Util.int2long( dim ), cache( type ) );
+	}
+
+	@Deprecated
+	public Img< T > create( final Supplier< T > typeSupplier, final long... dim )
+	{
+		return create( dim, cache( typeSupplier.get() ) );
+	}
+
+	@Deprecated
+	public Img< T > create( final Supplier< T > typeSupplier, final Dimensions dim )
+	{
+		return create( dim, cache( typeSupplier.get() ) );
+	}
+
+	@Deprecated
+	public Img< T > create( final Supplier< T > typeSupplier, final int[] dim )
+	{
+		return create( dim, cache( typeSupplier.get() ) );
+	}
+
+	/**
+	 * Makes a best effort to return a non-null type instance, caching the given
+	 * type if a cached type is not already in place.
+	 * <p>
+	 * Furthermore, if the cached type instance was previously null, the given
+	 * {@code type} becomes the cached instance. In this way, if the factory was
+	 * created using one of the deprecated typeless constructor signatures, but
+	 * then one of the deprecated {@code create} methods is called (i.e.: a
+	 * method which provides a type instance as an argument), the provided type
+	 * becomes the cached type instance so that subsequent invocations of the
+	 * typeless {@code create} methods will work as desired.
+	 * </p>
+	 *
+	 * @param type
+	 *            The type to return, and cache if needed.
+	 * @return A type instance matching the type of the factory. If the given
+	 *         {@code type} argument is non-null, it is returned. Otherwise, the
+	 *         cached type instance {@link #t} is returned.
+	 */
+	@Deprecated
+	protected T cache( final T type )
+	{
+		if ( t == null ) t = type;
+		return type == null ? t : type;
+	}
```

Comment:
This type cache in ImgFactory is unnecessary complicated.
See https://github.com/maarzt/imglib2/commit/ce7966f27abab694e4b261671fe1d4f270bf19e8